docs/dev/design/better-doc-urls-handling.rst

```diff
@@ -0,0 +1,417 @@
+     - /latest/manual/index.html
+     - subproject
+     -
+   * - /projects/subproject/en/latest/manual/index.html
+     - 404
+     - subproject
+     - The subproject is single version
+   * - /projects/subproject/404
+     - 404
+     - subproject
+     - The subproject exists, but the file doesn't
+   * - /projects/subproject/br/latest/manual/index.html
+     - /latest/br/latest/manual/index.html
+     - subproject
+     - The subproject has a ``br`` directory!
+   * - /projects/nothing/manual/index.html
+     - 404
+     - project
+     - The subproject doesn't exist
+   * - /en/latest/manual/index.html
+     - /latest/manual/index.html
+     - project
+     -
+   * - /404
+     - 404
+     - project
+     -
+
+Project with custom prefix
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- project (latest, 1.0)
+- subproject (latest, 1.0)
+
+``project`` has the ``prefix`` prefix, and ``sub`` subproject prefix.
+
+.. list-table:: 
+   :header-rows: 1
+
+   * - Request
+     - Requested file
+     - Current project
+     - Note
+   * - /en/latest/manual/index.html
+     - 404
+     - project
+     - The prefix doesn't match
+   * - /prefix/en/latest/manual/index.html
+     - /latest/manual/index.html
+     - project
+     -
+   * - /projects/subproject/en/latest/manual/index.html
+     - 404
+     - project
+     - The subproject prefix doesn't match
+   * - /sub/subproject/en/latest/manual/index.html
+     - /latest/manual/index.html
+     - subproject
+     -
+   * - /sub/nothing/en/latest/manual/index.html
+     - 404
+     - project
+     - The subproject doesn't exist
+
+Project with custom subproject prefix (empty)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- project (latest, 1.0)
+- subproject (latest, 1.0)
+
+``project`` has the ``/`` subproject prefix,
+this allow us to serve subprojects without using a prefix.
+
+.. list-table:: 
+   :header-rows: 1
+
+   * - Request
+     - Requested file
+     - Current project
+     - Note
+   * - /en/latest/manual/index.html
+     - /latest/manual/index.html
+     - project
+     -
+   * - /projects/subproject/en/latest/manual/index.html
+     - 404
+     - project
+     - The subproject prefix doesn't match
+   * - /subproject/en/latest/manual/index.html
+     - /latest/manual/index.html
+     - subproject
+     -
+   * - /nothing/en/latest/manual/index.html
+     - /latest/manual/index.html
+     - project
+     - The subproject/file doesn't exist
+
+Questions
+---------
+
+- When using custom URLs,
+  should we support changing the URLs
+  that aren't related to doc serving?
+
+  These are:
+
+  - Health check
+  - Proxied APIs
+  - robots and sitemap
+  - The ``page`` redirect
+
+  I'd say we shouldn't, I can't think of a reason why this should be supported.
+
+- Should we use the urlconf from the subproject when processing it?
+  This is an URL like ``/projects/subproject/custom/prefix/en/latest/index.html``.
+
+  I don't think that's useful, but it should be easy to support if needed.
```

Comment:
We need to think more about this. I can see users who might want to have eg. just a verison and not a translation exposed on the subproject:

`/projects/subproject/1.0/index.html` which would serve `subproject/$lang/1.0/index.html` or similar. Would we take the custom URLconf for the subproject from the canonical project? I don't know..

Reply:
Hrm, perhaps. I guess it would work similarly to single version projects, and just remove it from the URL?

Reply:
But them, we need to restrict those projects from having translations, otherwise we will try to serve them from the domain of the main translation, and that link would be impossible to generate.

Reply:
Instead of adding extra attributes to the project and telling people to enable/disable them, we can define the type of project via its URL pattern.

* `/<lang>/<version>/<filename>/`
  * multilanguage
  * multiversion
* `/<lang>/<filename>`
  * multilanguage
  * single version
* `/<version>/<filename>/`
  * multiversion
  * single language

If the user tells us what the URLs will look like, we know exactly what type of project it is and how to parse and generate the URLs for them.